File: LBExpandFriendList/LBExpandFriendList/折叠的表示图/LBExpandController.m
//
//  LBExpandController.m
//  LBExpandFriendList
//
//  Created by yunmei on 2018/1/19.
//  Copyright © 2018年 yunmei. All rights reserved.
//

#import "LBExpandController.h"
#import "LBExpandCell.h"
@interface LBExpandController ()<UITableViewDataSource,UITableViewDelegate>
@property (nonatomic, strong)UITableView *lbexpandTableView;
@property (nonatomic, strong)NSMutableArray *dataSource;
@property (nonatomic, strong)NSMutableArray *sectionArray;
@property (nonatomic, strong)NSMutableArray *stateArray;

@end

@implementation LBExpandController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"可以折叠的表示图";
    self.view.backgroundColor = [UIColor whiteColor];
    
    /// 添加表示图数据
    [self addData];
    /// 添加表示图
    [self addTableView];
}
-(void)addTableView{
    _lbexpandTableView = [[UITableView alloc]initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height) style:UITableViewStylePlain];
    _lbexpandTableView.dataSource = self;
    _lbexpandTableView.delegate =  self;
    _lbexpandTableView.tableFooterView = [UIView new];
    [_lbexpandTableView registerNib:[UINib nibWithNibName:@"ExpandCell" bundle:nil] forCellReuseIdentifier:@"cell"];
    [self.view addSubview:_lbexpandTableView];
}
- (void)addData{
    _sectionArray  = [NSMutableArray arrayWithObjects:@"一级分类",
                     @"一级分类",
                     @"一级分类",
                     @"一级分类",nil];
    
    NSArray *one = @[@"二级分类",@"二级分类",@"二级分类"];
    NSArray *two = @[@"二级分类",@"二级分类",@"二级分类"];
    NSArray *three = @[@"二级分类",@"二级分类",@"二级分类"];
    NSArray *four = @[@"二级分类",@"二级分类",@"二级分类",@"二级分类"];
    
    _dataSource = [NSMutableArray arrayWithObjects:one,two,three,four, nil];
    _stateArray = [NSMutableArray array];
    
    for (int i = 0; i < _dataSource.count; i++)
    {
        //所有的分区都是闭合
        [_stateArray addObject:@"0"];
    }
}
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return _dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    LBExpandCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
//    cell.listLabel.textAlignment = NSTextAlignmentLeft;
//    cell.listLabel.text = _dataSource[indexPath.section][indexPath.row];
    cell.backgroundColor = [UIColor whiteColor];
    cell.selectionStyle = UITableViewCellSelectionStyleDefault;
    cell.contentView.backgroundColor = [UIColor whiteColor];
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return _sectionArray[section];
}


@end

File: LBSwiftExpandFriendList/LBSwiftExpandFriendList/可折叠的表示图/LBExpandController.swift
//
//  LBExpandController.swift
//  LBSwiftExpandFriendList
//
//  Created by yunmei on 2018/1/19.
//  Copyright © 2018年 liubo. All rights reserved.
//

import UIKit

class LBExpandController: UIViewController {
    /// 所有组的数据
    lazy var dataSourceArray = NSMutableArray()
    /// 组标题
    lazy var sectionTitleArray = { () -> [String] in
        let titleArray = ["一级分类","一级分类","一级分类","一级分类","一级分类","一级分类","一级分类","一级分类","一级分类","一级分类","一级分类"]
        return titleArray
    }()
    /// 每组展开状态
    lazy var sectionStateArray = NSMutableArray()
    /// 表示图
    lazy var lbexpandTableView:LBExpandTableView = {
        let expandTableView = LBExpandTableView.init(frame: UIScreen.main.bounds, style: .grouped)
        return expandTableView
    }()
    override func viewDidLoad() {
        super.viewDidLoad()
        title = "可以折叠的表示图"
        view.backgroundColor = UIColor.white
        /// 添加表示图数据
        addData()
    }

}
extension LBExpandController{
    func addData(){
        /// 下面是数据测试实例
        let zero = ["二级分类"];
        let one = ["二级分类","二级分类"];
        let two = ["二级分类","二级分类","二级分类"];
        let three = ["二级分类","二级分类","二级分类","二级分类"];
        let four = ["二级分类","二级分类","二级分类","二级分类","二级分类"];
        let five = ["二级分类","二级分类","二级分类","二级分类","二级分类","二级分类"];
        let six = ["二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类"];
        let seven = ["二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类"];
        let eight = ["二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类"];
        let nine = ["二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类","二级分类"];
        dataSourceArray.addObjects(from: [zero,one,two,three,four,five,six,seven,eight,nine])
        
        for _ in dataSourceArray {
            /// 默认所有的组都是闭合的
            sectionStateArray.add("0")
        }
        lbexpandTableView.dataSourceArray = dataSourceArray as! [[String]]
        lbexpandTableView.sectionStateArray = sectionStateArray
//        lbexpandTableView.sectionTitleArray = sectionTitleArray as! NSMutableArray
    }
}
